### PR TITLE
feat(parameters): migrate AppConfig to new APIs due to API deprecation

### LINKED ISSUE #1506: Maintenance: Change parameters appconfig utility API due to GetConfiguration deprecation
### Summary

The current API will be deprecated in the near future according to AWS AppConfig team.
See [this](https://docs.aws.amazon.com/appconfig/2019-10-09/APIReference/API_GetConfiguration.html)
See https://github.com/awslabs/aws-lambda-powertools-python/blob/develop/aws_lambda_powertools/utilities/parameters/appconfig.py

### Why is this needed?

GetConfiguration - this API action has been deprecated. Calls to receive configuration data should use the [StartConfigurationSession](https://docs.aws.amazon.com/appconfig/2019-10-09/APIReference/API_appconfigdata_StartConfigurationSession.html) and [GetLatestConfiguration](https://docs.aws.amazon.com/appconfig/2019-10-09/APIReference/API_appconfigdata_GetLatestConfiguration.html) APIs instead.

### Which area does this relate to?

Parameters

### Solution

Use new APIs, hopefully in a non breaking manner.

### Acknowledgment

- [X] This request meets [Lambda Powertools Tenets](https://awslabs.github.io/aws-lambda-powertools-python/latest/#tenets)
- [X] Should this be considered in other Lambda Powertools languages? i.e. [Java](https://github.com/awslabs/aws-lambda-powertools-java/), [TypeScript](https://github.com/awslabs/aws-lambda-powertools-typescript/)

### PR DESCRIPTION
**Pull request by leandrodamascena**
**Issue number:** #1506 

### Changes

Due to [Deprecation of the AppConfig API](https://docs.aws.amazon.com/appconfig/2019-10-09/APIReference/API_GetConfiguration.html), we needed to change the SDK client to call the new APIs.

We had a long discussion in this issue #1506 before applying these changes.

### User experience

No changes are required to Lambda functions that use `Feature Flags and AppConfig Parameter utility`, but users must add `appconfig:GetLatestConfiguration` and `appconfig:StartConfigurationSession` IAM permissions for this to work in their Lambdas.

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x ] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [x ] I have performed a self-review of this change
* [x ] Changes have been tested
* [x ] Changes are documented
* [x ] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

-----
[View rendered docs/upgrade.md](https://github.com/leandrodamascena/aws-lambda-powertools-python/blob/feat/new-appconfig-apis/docs/upgrade.md)

### CODE REVIEW

File: tests/e2e/parameters/conftest.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import pytest
+
+from tests.e2e.parameters.infrastructure import ParametersStack
```
Comment (leandrodamascena) [inline tests/e2e/parameters/conftest.py:3]: I don't know what's going on, but I couldn't add the other commits with more e2e tests. I'm fixing it locally and will add it soon.

File: docs/upgrade.md [hunk 2]
```diff
@@ -154,3 +155,9 @@ Prior to this change, the Idempotency key was generated using only the caller fu
 After this change, the key is generated using the `module name` + `qualified function name` + `idempotency key` (e.g: `app.classExample.function#app.handler#282e83393862a613b612c00283fef4c8`).
 
 Using qualified names prevents distinct functions with the same name to contend for the same Idempotency key.
+
+## Feature Flags and AppConfig Parameter utility
```
Comment (leandrodamascena) [inline docs/upgrade.md:159]: Hi @heitorlessa! I need help to improve the documentation. Maybe adding best practices using all parameters under the same configuration?

File: aws_lambda_powertools/utilities/parameters/appconfig.py [hunk 3]
```diff
@@ -100,16 +100,23 @@ def _get(self, name: str, **sdk_options) -> str:
...
+            sdk_options["ApplicationIdentifier"] = self.application
+            sdk_options["EnvironmentIdentifier"] = self.environment
+            response_configuration = self.client.start_configuration_session(**sdk_options)
+            self.next_call = response_configuration["InitialConfigurationToken"]
 
-        sdk_options["Configuration"] = name
-        sdk_options["Application"] = self.application
-        sdk_options["Environment"] = self.environment
-        sdk_options["ClientId"] = CLIENT_ID
+        response = self.client.get_latest_configuration(ConfigurationToken=self.next_call)
```
Comment (ran-isenberg) [inline aws_lambda_powertools/utilities/parameters/appconfig.py]: arent you supposed to add here the sdk_options? let's say i want to config from a different region or some other config

Comment (leandrodamascena) [reply aws_lambda_powertools/utilities/parameters/appconfig.py]: Thanks for pointing out this @ran-isenberg. Configuration like region, retries, credentials and others are passed to the constructor and not to the API calls, that is, if you check [this line](https://github.com/awslabs/aws-lambda-powertools-python/pull/1553/files#diff-5576360015200fe47fe1113f8fcee3d78e9fdb6059051e7663d220c7dadcf78bR82) you see any additional configuration is passed to the constructor.

I created this fictional code to make sure we are not missing anything
image
image
image

I think you mentioned a situation where a customer needs to pass [additional parameters to the SDK](https://awslabs.github.io/aws-lambda-powertools-python/latest/utilities/parameters/#passing-additional-sdk-arguments) and it makes sense, but I don't think that's the situation here. The method [get_latest_configuration](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/appconfigdata.html#AppConfigData.Client.get_latest_configuration) only supports the `ConfigurationToken` argument and if in the future it supports more than this argument, we need to review the impact of a possible change.

Comment (ran-isenberg) [reply aws_lambda_powertools/utilities/parameters/appconfig.py]: yes, my bad! I looked at the wrong api doc. sorry about that.

**Comment by leandrodamascena** [comment]
Giving feedback here. 
I need to add more information in the doc files, but we hope to review and merge this PR later this week.

**Comment by heitorlessa** [comment]
Looking into this now [apologies for the delay, re:Invent deadlines are tighter this year]

File: aws_lambda_powertools/utilities/parameters/appconfig.py [hunk 4]
```diff
@@ -100,16 +100,26 @@ def _get(self, name: str, **sdk_options) -> str:
         name: str
             Name of the configuration
         sdk_options: dict, optional
-            Dictionary of options that will be passed to the client's get_configuration API call
+            Dictionary of options that will be passed to the client's start_configuration_session API call
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/parameters/appconfig.py]:
```suggestion
            SDK options to propagate to `start_configuration_session` API call
```

Fixing older and suboptimal wording (when this was introduced)

File: aws_lambda_powertools/utilities/parameters/appconfig.py [hunk 5]
```diff
@@ -91,6 +88,9 @@ def __init__(
         self.environment = environment
         self.current_version = ""
 
+        self.next_call = ""
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/parameters/appconfig.py]:
```suggestion
        self._next_token = ""
```

Unless we're planning to expose the next token to customers, let's make it private.

File: aws_lambda_powertools/utilities/parameters/appconfig.py [hunk 6]
```diff
@@ -145,7 +155,7 @@ def get_app_config(
     max_age: int
         Maximum age of the cached value
     sdk_options: dict, optional
-        Dictionary of options that will be passed to the boto client get_configuration API call
+        Dictionary of options that will be passed to the boto client start_configuration_session API call
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/parameters/appconfig.py]:
```suggestion
        SDK options to propagate to `start_configuration_session` API call
```

File: docs/upgrade.md [hunk 7]
```diff
@@ -13,6 +13,7 @@ Changes at a glance:
 * The API for **event handler's `Response`** has minor changes to support multi value headers and cookies.
 * The **legacy SQS batch processor** was removed.
 * The **Idempotency key** format changed slightly, invalidating all the existing cached results.
+* The **Feature Flags and AppConfig Parameter utility** API calls have changed and you must add new IAM permissions to Lambda.
```
Comment (heitorlessa) [inline docs/upgrade.md]:
```suggestion
* The **Feature Flags and AppConfig Parameter utility** API calls have changed and you must update your IAM permissions.
```

File: docs/upgrade.md (hunk 2 above)
Comment (heitorlessa) [reply docs/upgrade.md:159]: Sure. We should add an `Info` Admonition in the Parameters and Feature Flags docs.

File: docs/upgrade.md [hunk 8]
```diff
@@ -154,3 +155,9 @@ Prior to this change, the Idempotency key was generated using only the caller fu
 After this change, the key is generated using the `module name` + `qualified function name` + `idempotency key` (e.g: `app.classExample.function#app.handler#282e83393862a613b612c00283fef4c8`).
 
 Using qualified names prevents distinct functions with the same name to contend for the same Idempotency key.
+
+## Feature Flags and AppConfig Parameter utility
+
+The current API is [deprecated](https://docs.aws.amazon.com/appconfig/2019-10-09/APIReference/API_GetConfiguration.html) for new accounts and we had to change this.
+
+No changes are required to your code, but you must add `appconfig:GetLatestConfiguration` and `appconfig:StartConfigurationSession` IAM permissions for this to work in your Lambdas.
```
Comment (heitorlessa) [inline docs/upgrade.md]:
```suggestion
AWS AppConfig deprecated the current API (GetConfiguration) - [more details here](https://github.com/awslabs/aws-lambda-powertools-python/issues/1506#issuecomment-1266645884).

You must update your IAM permissions to allow `appconfig:GetLatestConfiguration` and `appconfig:StartConfigurationSession`. There are no code changes required.
```

File: docs/utilities/feature_flags.md [hunk 9]
```diff
@@ -27,6 +27,10 @@ If you want to learn more about feature flags, their variations and trade-offs,
 * [Feature Toggles (aka Feature Flags) - Pete Hodgson](https://martinfowler.com/articles/feature-toggles.html)
 * [AWS Lambda Feature Toggles Made Simple - Ran Isenberg](https://isenberg-ran.medium.com/aws-lambda-feature-toggles-made-simple-580b0c444233)
 * [Feature Flags Getting Started - CloudBees](https://www.cloudbees.com/blog/ultimate-feature-flag-guide)
+* [Best Practices for validating AWS AppConfig Feature Flags and Configuration Data](https://aws.amazon.com/pt/blogs/mt/best-practices-for-validating-aws-appconfig-feature-flags-and-configuration-data/)
```
Comment (heitorlessa) [inline docs/utilities/feature_flags.md]:
```suggestion
```

Not applicable here. You also accidentally shared a `pt` domain

File: docs/utilities/feature_flags.md [hunk 10]
```diff
@@ -27,6 +27,10 @@ If you want to learn more about feature flags, their variations and trade-offs,
...
+
+???+ note
+    Optimize your Lambda execution time by putting your feature settings in a single [Configuration](https://docs.aws.amazon.com/appconfig/latest/userguide/appconfig-creating-configuration-and-profile.html).
```
Comment (heitorlessa) [inline docs/utilities/feature_flags.md]:
```suggestion
    AWS AppConfig requires two API calls to fetch configuration for the first time. You can improve latency  by consolidating your feature settings in a single [Configuration](https://docs.aws.amazon.com/appconfig/latest/userguide/appconfig-creating-configuration-and-profile.html).
```

In this case, a tip is to share why they should continue reading this note - Challenge first, recommended step afterward. If the challenge doesn't concern them, they can stop reading immediately. This optimizes for cognitive load.

File: docs/utilities/feature_flags.md [hunk 11]
```diff
@@ -38,7 +42,7 @@ If you want to learn more about feature flags, their variations and trade-offs,
 
 ### IAM Permissions
 
-Your Lambda function must have `appconfig:GetConfiguration` IAM permission in order to fetch configuration from AWS AppConfig.
+Your Lambda function must have `appconfig:GetLatestConfiguration` and `appconfig:StartConfigurationSession` IAM permissions in order to fetch configuration from AWS AppConfig.
```
Comment (heitorlessa) [inline docs/utilities/feature_flags.md]:
```suggestion
Your Lambda function IAM Role must have `appconfig:GetLatestConfiguration` and `appconfig:StartConfigurationSession` IAM permissions before using this feature.
```

fixing older wording

File: tests/functional/test_utilities_parameters.py [hunk 12]
```diff
@@ -1639,14 +1639,22 @@ def test_appconf_provider_get_configuration_json_content_type(mock_name, config)
     encoded_message = json.dumps(mock_body_json).encode("utf-8")
     mock_value = StreamingBody(BytesIO(encoded_message), len(encoded_message))
 
-    # Stub the boto3 client
     stubber = stub.Stubber(provider.client)
-    response = {"Content": mock_value, "ConfigurationVersion": "1", "ContentType": "application/json"}
-    stubber.add_response("get_configuration", response)
+    response_start = {"InitialConfigurationToken": "initial_token"}
+    stubber.add_response("start_configuration_session", response_start)
```
Comment (heitorlessa) [inline tests/functional/test_utilities_parameters.py]:
```suggestion
    response_start_config_session = {"InitialConfigurationToken": "initial_token"}
    stubber.add_response("start_configuration_session", response_start_config_session)
```

easier to read which response belongs to which API

File: tests/functional/test_utilities_parameters.py [hunk 13]
```diff
@@ -1639,14 +1639,22 @@ def test_appconf_provider_get_configuration_json_content_type(mock_name, config)
...
+
+    response_latest = {
+        "Configuration": mock_value,
+        "NextPollConfigurationToken": "initial_token",
+        "ContentType": "application/json",
+    }
+    stubber.add_response("get_latest_configuration", response_latest)
```
Comment (heitorlessa) [inline tests/functional/test_utilities_parameters.py]:
```suggestion
    response_get_latest_config = {
        "Configuration": mock_value,
        "NextPollConfigurationToken": "initial_token",
        "ContentType": "application/json",
    }
    stubber.add_response("get_latest_configuration", response_get_latest_config)
```

easier to read which response belongs to which API

File: tests/functional/test_utilities_parameters.py [hunk 14]
```diff
@@ -1670,14 +1678,22 @@ def test_appconf_provider_get_configuration_json_content_type_with_custom_client
     encoded_message = json.dumps(mock_body_json).encode("utf-8")
     mock_value = StreamingBody(BytesIO(encoded_message), len(encoded_message))
 
-    # Stub the boto3 client
     stubber = stub.Stubber(provider.client)
-    response = {"Content": mock_value, "ConfigurationVersion": "1", "ContentType": "application/json"}
-    stubber.add_response("get_configuration", response)
+    response_start = {"InitialConfigurationToken": "initial_token"}
+    stubber.add_response("start_configuration_session", response_start)
```
Comment (heitorlessa) [inline tests/functional/test_utilities_parameters.py]:
```suggestion
    response_start_config_session = {"InitialConfigurationToken": "initial_token"}
    stubber.add_response("start_configuration_session", response_start_config_session)
```

File: tests/functional/test_utilities_parameters.py [hunk 15]
```diff
@@ -1670,14 +1678,22 @@ def test_appconf_provider_get_configuration_json_content_type_with_custom_client
...
+
+    response_latest = {
+        "Configuration": mock_value,
+        "NextPollConfigurationToken": "initial_token",
+        "ContentType": "application/json",
+    }
+    stubber.add_response("get_latest_configuration", response_latest)
```
Comment (heitorlessa) [inline tests/functional/test_utilities_parameters.py]:
```suggestion
    response_get_latest_config = {
        "Configuration": mock_value,
        "NextPollConfigurationToken": "initial_token",
        "ContentType": "application/json",
    }
    stubber.add_response("get_latest_configuration", response_get_latest_config)
```

File: tests/functional/test_utilities_parameters.py [hunk 16]
```diff
@@ -1699,10 +1715,16 @@ def test_appconf_provider_get_configuration_no_transform(mock_name, config):
...
```
Comment (heitorlessa) [inline tests/functional/test_utilities_parameters.py]:
```suggestion
    response_get_latest_config = {
        "Configuration": mock_value,
        "NextPollConfigurationToken": "initial_token",
        "ContentType": "application/json",
    }
    stubber.add_response("get_latest_configuration", response_get_latest_config)
```

File: tests/functional/test_utilities_parameters.py [hunk 17]
```diff
@@ -1699,10 +1715,16 @@ def test_appconf_provider_get_configuration_no_transform(mock_name, config):
     encoded_message = json.dumps(mock_body_json).encode("utf-8")
     mock_value = StreamingBody(BytesIO(encoded_message), len(encoded_message))
 
-    # Stub the boto3 client
     stubber = stub.Stubber(provider.client)
-    response = {"Content": mock_value, "ConfigurationVersion": "1", "ContentType": "application/json"}
-    stubber.add_response("get_configuration", response)
+    response_start = {"InitialConfigurationToken": "initial_token"}
+    stubber.add_response("start_configuration_session", response_start)
```
Comment (heitorlessa) [inline tests/functional/test_utilities_parameters.py]:
```suggestion
    response_start_config_session = {"InitialConfigurationToken": "initial_token"}
    stubber.add_response("start_configuration_session", response_start_config_session)
```

File: tests/e2e/parameters/test_ssm.py [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+def test_simple_parameter_string(parameter_string_handler_fn_arn: str, parameter_string_value: str):
+    # GIVEN
+    expected_return = json.dumps({"value": parameter_string_value})
+
+    # WHEN
+    parameter_execution, _ = data_fetcher.get_lambda_response(lambda_arn=parameter_string_handler_fn_arn)
+    parameter_value = parameter_execution["Payload"].read().decode("utf-8")
+
+    # THEN
+    assert parameter_value == expected_return
```
Comment (heitorlessa) [inline tests/e2e/parameters/test_ssm.py]: could you split this into a separate PR? If we need to revert this AppConfig PR, we won't affect SSM E2E. It also helps us to have a better focus on SSM `with_decryption=True` too.

File: tests/e2e/parameters/test_appconfig.py [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+@pytest.fixture
+def parameter_appconfig_freeform_handler_fn(infrastructure: dict) -> str:
+    return infrastructure.get("ParameterAppconfigFreeformHandler", "")
+
+
+@pytest.fixture
+def parameter_appconfig_freeform_value(infrastructure: dict) -> str:
+    return infrastructure.get("AppConfigConfigurationValue", "")
+
+
```
Comment (heitorlessa) [inline tests/e2e/parameters/test_appconfig.py:22]: Should we have a test for FeatureFlags ConfigurationProfile type too?

I don't recall if we have full control on the content of that type in AppConfig (as in, add our expected schema). If we do, it'd be contribute to a more comprehensive coverage to detect this feature will work in both scenarios.

PS: The fetching will be exactly the same, so your Lambda function can be reused (remove `freeform` from the handlers name)

File: tests/e2e/parameters/infrastructure.py [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+        cfn_environment = appconfig.CfnEnvironment(
+            self.stack, "appconfig-env", application_id=cfn_application.ref, name="enve2e", description="enve2e"
+        )
+        CfnOutput(self.stack, "AppConfigEnvironment", value=cfn_environment.name)
+
+        cfn_deployment_strategy = appconfig.CfnDeploymentStrategy(
+            self.stack,
+            "appconfig-deployment-strategy",
+            deployment_duration_in_minutes=0,
+            final_bake_time_in_minutes=0,
```
Comment (heitorlessa) [inline tests/e2e/parameters/infrastructure.py:38]: Superb!!

File: tests/e2e/parameters/infrastructure.py [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+            growth_factor=100,
+            name="deploymente2e",
```
Comment (heitorlessa) [inline tests/e2e/parameters/infrastructure.py]: I'm 98% certain this won't be a problem, but I wonder if AppConfig allows multiple Deployment Strategies with the same name in the event of parallel tests running (3.7-3.9).

If we do see any CI error, we can easily use our builder to generate random values.

File: tests/e2e/parameters/infrastructure.py [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+        )
+
+        parameter.grant_read(function)
+
+        CfnOutput(self.stack, "ParameterString", value=parameter.parameter_name)
+        CfnOutput(self.stack, "ParameterStringValue", value=parameter.string_value)
+
+    def _create_app_config(self, function: Function):
+
+        cfn_application = appconfig.CfnApplication(self.stack, id="appconfig-app", name="appe2e", description="appe2e")
```
Comment (heitorlessa) [inline tests/e2e/parameters/infrastructure.py]: Same as my other comment - do you think we can run into an issue with the name being hard coded?

e.g., parallel tests, tests that failed due to networking and left orphan resources, etc.

File: tests/e2e/parameters/infrastructure.py [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+
+class ParametersStack(BaseInfrastructure):
+    def create_resources(self):
+        functions = self.create_lambda_functions()
+        self._create_parameter_string(function=functions["ParameterStringHandler"])
+        self._create_app_config(function=functions["ParameterAppconfigFreeformHandler"])
+
+    def _create_parameter_string(self, function: Function):
+        parameter = ssm.StringParameter(
+            self.stack, id="string_parameter", parameter_name="sample_string", string_value="Lambda Powertools"
```
Comment (heitorlessa) [inline tests/e2e/parameters/infrastructure.py]: We need to split SSM into a separate PR. When you do, we need to ensure the parameter name is unique otherwise it'll fail subsequent tests running in parallel

**Review by heitorlessa — CHANGES_REQUESTED** [review]
Thanks a lot @leandrodamascena for working on this, truly!

I've made some wording suggestions, and some concerns on E2E tests:

* Separate SSM from this PR for a clean cut and revert if we need
* Some resources have names hardcoded, this might bite us back in parallel testing or network failures (e.g., SSM will fail subsequent stack if a param name already exists)

File: tests/e2e/parameters/test_ssm.py (hunk 18 above)
Comment (leandrodamascena) [reply tests/e2e/parameters/test_ssm.py]: I agree 100% @heitorlessa! I removed it and I will raise a new issue for this.

File: tests/e2e/parameters/infrastructure.py (hunk 23 above)
Comment (leandrodamascena) [reply tests/e2e/parameters/infrastructure.py]: I agree 100% @heitorlessa! I removed it and I will raise a new issue for this.

File: tests/e2e/parameters/infrastructure.py (hunk 22 above)
Comment (leandrodamascena) [reply tests/e2e/parameters/infrastructure.py]: @heitorlessa thanks for bringing all these insights! This is very important so that I don't make these mistakes again in the future. :heart:

File: tests/e2e/parameters/test_appconfig.py (hunk 19 above)
Comment (leandrodamascena) [reply tests/e2e/parameters/test_appconfig.py:22]: @heitorlessa currently we don't support [AWS.AppConfig.FeatureFlags](https://docs.aws.amazon.com/appconfig/latest/userguide/appconfig-creating-configuration-and-profile.html#appconfig-type-reference-feature-flags) profile and I added a info in [documentation](https://github.com/awslabs/aws-lambda-powertools-python/pull/1553/files#diff-0e8fe1a0c18945a2f28b5a87df6c635fbbbfc7d87e44305a0978e41db488024bR8)

**Comment by leandrodamascena** [comment]
> Thanks a lot @leandrodamascena for working on this, truly!
> 
> I've made some wording suggestions, and some concerns on E2E tests:
> 
> * Separate SSM from this PR for a clean cut and revert if we need
> * Some resources have names hardcoded, this might bite us back in parallel testing or network failures (e.g., SSM will fail subsequent stack if a param name already exists)

Heiiiiii @heitorlessa! I've addressed all the feedback and I think we're ready to merge.

Thank you so much for reviewing this, every time you give me WONDERFUL feedback to think about UX and see beyond the code. :heart: 

image

File: docs/utilities/feature_flags.md [hunk 24]
```diff
@@ -3,11 +3,12 @@ title: Feature flags
...
 ---
 
-???+ note
-    This is currently in Beta, as we might change Store parameters in the next release.
-
 The feature flags utility provides a simple rule engine to define when one or multiple features should be enabled depending on the input.
 
+???+ info
+    We currently only support AppConfig using [freeform configuration profile](https://docs.aws.amazon.com/appconfig/latest/userguide/appconfig-creating-configuration-and-profile.html#appconfig-creating-configuration-and-profile-free-form-configurations).
+    We are planning to add support for the [feature flag configuration profile](https://docs.aws.amazon.com/appconfig/latest/userguide/appconfig-creating-configuration-and-profile.html#appconfig-creating-configuration-and-profile-feature-flags) in future releases.
```
Comment (heitorlessa) [inline docs/utilities/feature_flags.md]:
```suggestion
```

In this case, it's best to wait for customer demand via feature request (well known format). If we add that we're planning to add support now, we'll accidentally commit and prioritize over existing backlog items other customers asked.

File: tests/e2e/parameters/infrastructure.py [hunk 25]
```diff
@@ -3,34 +3,31 @@
...
-        CfnOutput(self.stack, "ParameterString", value=parameter.parameter_name)
-        CfnOutput(self.stack, "ParameterStringValue", value=parameter.string_value)
-
     def _create_app_config(self, function: Function):
 
-        cfn_application = appconfig.CfnApplication(self.stack, id="appconfig-app", name="appe2e", description="appe2e")
+        service_name = build_service_name()
+
+        cfn_application = appconfig.CfnApplication(
+            self.stack, id="appconfig-app", name=f"appe2e{service_name}", description="appe2e"
```
Comment (heitorlessa) [inline tests/e2e/parameters/infrastructure.py]:
```suggestion
            self.stack, id="appconfig-app", name=f"powertools-e2e-{service_name}", description="Lambda Powertools End-to-End testing for AppConfig"
```

making it more explicit in the event of dangling resources caused by network split

File: tests/e2e/parameters/infrastructure.py [hunk 26]
```diff
@@ -3,34 +3,31 @@
...
+        )
         CfnOutput(self.stack, "AppConfigApplication", value=cfn_application.name)
 
         cfn_environment = appconfig.CfnEnvironment(
-            self.stack, "appconfig-env", application_id=cfn_application.ref, name="enve2e", description="enve2e"
+            self.stack,
+            "appconfig-env",
+            application_id=cfn_application.ref,
+            name=f"enve2e{service_name}",
```
Comment (heitorlessa) [inline tests/e2e/parameters/infrastructure.py]:
```suggestion
            name=f"powertools-e2e{service_name}",
```

File: tests/e2e/parameters/infrastructure.py [hunk 27]
```diff
@@ -3,34 +3,31 @@
...
+            description="enve2e",
```
Comment (heitorlessa) [inline tests/e2e/parameters/infrastructure.py]:
```suggestion
            description="Lambda Powertools End-to-End testing environment",
```

**Review by heitorlessa — APPROVED** [review]
THANK YOU! I'm glad we're completing this and we're ready for the grooming part before launching V2 now.

Last suggestions I'm committing myself is on being more explicit on AppConfig Config/Env Name and Description. If a network split happens, this will leave dangling resources making it difficult to trace back to what they are for (`appe2e` vs `powertools-e2e-x`).
